include/picongpu/plugins/binning/WriteHist.hpp

```diff
@@ -0,0 +1,171 @@
+/* Copyright 2023 Tapish Narwal
+ *
+ * This file is part of PIConGPU.
+ *
+ * PIConGPU is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * PIConGPU is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#if !ENABLE_OPENPMD
+#    error The activated Binning plugin requires openPMD-api.
+#endif
+
+#pragma once
+
+#include "picongpu/plugins/binning/UnitConversion.hpp"
+#include "picongpu/plugins/binning/utility.hpp"
+#include "picongpu/plugins/common/openPMDDefaultExtension.hpp"
+#include "picongpu/plugins/common/stringHelpers.hpp"
+
+#include <openPMD/openPMD.hpp>
+
+namespace picongpu
+{
+    namespace plugins::binning
+    {
+        struct WriteOpenPMDParams
+        {
+            std::string const& extension;
+            std::string const& jsonConfig;
+        };
+
+        /**
+         * Write the N-D histogram with units
+         * Write axis bin edges with units
+         */
+        class WriteHist
+        {
+        private:
+            std::string openPMDSuffix = "_%T." + openPMD::getDefaultExtension(openPMD::ExtensionPreference::HDF5);
+
+        public:
+            template<typename T_Type, typename T_BinningData>
+            void operator()(
+                HostBuffer<T_Type, 1u>& hReducedBuffer,
+                T_BinningData binningData,
+                std::string const& dir,
+                // WriteOpenPMDParams const& params,
+                std::array<double, 7> outputUnits,
+                const uint32_t currentStep)
+            {
+                using Type = T_Type;
+
+                // auto const& [extension, jsonConfig] = params;
+                auto const& extension = openPMDSuffix;
+                std::ostringstream filename;
+                if(std::any_of(extension.begin(), extension.end(), [](char const c) { return c == '.'; }))
+                {
+                    filename << binningData.binnerOutputName << extension;
+                }
+                else
+                {
+                    filename << binningData.binnerOutputName << '.' << extension;
+                }
+
+                auto series = ::openPMD::Series(dir + '/' + filename.str(), ::openPMD::Access::CREATE);
```

Comment:
Not necessary to do in this PR, but if at some point we want more than file-based encoding (one file per iteration), we should consider leaving the `Series` open between calls to this plugin.